blogposts/2022/introducting-migrator-service.md

```diff
@@ -0,0 +1,149 @@
+
+On the customer side, the resolution is very inelegant. A new container will come online and attempt to migrate the database. When this attempt fails, the database is marked as dirty and the container will restart. New instances will see the dirty flag set and also restart. Site-administrators will notice rapidly restarting containers (that, or a wonky Sourcegraph experience) and eventually might find a log message asking them to contact support for assistance.
+
+In our Cloud environment, Sourcegraph engineers are subjected to the same manual process.
+
+Requiring site administrators to directly modify database contents is not a distribution practice we should encourage. The frequency of these errors and the rising burden put on users and application/product engineers made this a pressing issue to resolve.
+
+## Identifying (a) contributing factor
+
+In our original effort to erradicate this class of error, we incorrectly attributed the Docker container health check period as the primary cause. Our intuition was built from two observations:
+
+1. We never see this issue in local development which are usually operate over relatively small sets of data.
+2. We see this issue frequently in our Cloud environment, but exclusively on objects over a particular (but undetermined) size, such as our repository or code intelligence data tables.
+
+We believed that the issue occurred if and only if the set of migrations running on application startup could not finish within the parent container's health check timeout. If the migration took longer to perform than the configured timeout, Kubernetes would declare the pod dead and restart it. Killing a pod _during_ an active migration marks the database as dirty until the site administrator manually resolves the issue.
+
+And this is exactly the behavior we see.
+
+Our original solution, simply "decoupling" the database from application upgrades, targeted the Kubernetes health check as the primary problem. By performing the migrations separately from the application deployment, the migration process no longer has to complete within a configured healthcheck timeout of a downstream application. By having a separate component performing the migrations, and by moving away from our dependency on a [third-party](https://github.com/golang-migrate/migrate) migration library, we can better inspect the health of the migrator and the progress of its migration attempts.
+
+The first draft of the `migrator` service kept all other behavior the same. We explicitly kept the new migration behavior as close to the previous migration behavior as possible - just running in a different container - to avoid turning too many knobs at once. Part of the behavior we explicitly did _not_ change was how we handle concurrent migration attempts.
```

Comment:
```suggestion
The first draft of the `migrator` service kept all other behavior the same. We explicitly kept the new migration behavior as close to the previous migration behavior as possible–just running in a different container–to avoid turning too many knobs at once. Part of the behavior we explicitly did _not_ change was how we handle concurrent migration attempts.
```